src/sinks/statsd.rs

```diff
@@ -329,75 +328,65 @@ mod test {
-            .map_err(drop)
-            .and_then(|_| sender);
-
-        let (tx, rx) = mpsc::channel(1);
-
-        let receiver = Box::new(
-            future::lazy(|| {
-                let socket = UdpSocket::bind(&default_address()).unwrap();
-                future::ok(socket)
-            })
-            .and_then(|socket| {
+        rt.block_on_std(async move {
+            let config = StatsdSinkConfig {
+                namespace: "vector".into(),
+                address: default_address(),
+                batch: BatchConfig {
+                    max_bytes: Some(512),
+                    timeout_secs: Some(1),
+                    ..Default::default()
+                },
+            };
+            let sink = StatsdSvc::new(config, Acker::Null).unwrap();
+
+            let events = vec![
+                Event::Metric(Metric {
+                    name: "counter".to_owned(),
+                    timestamp: None,
+                    tags: Some(tags()),
+                    kind: MetricKind::Incremental,
+                    value: MetricValue::Counter { value: 1.5 },
+                }),
+                Event::Metric(Metric {
+                    name: "histogram".to_owned(),
+                    timestamp: None,
+                    tags: None,
+                    kind: MetricKind::Incremental,
+                    value: MetricValue::Distribution {
+                        values: vec![2.0],
+                        sample_rates: vec![100],
+                    },
+                }),
+            ];
+            let (tx, rx) = mpsc::channel(1);
+
+            tokio::spawn(async move {
+                let socket = UdpSocket::bind(default_address()).await.unwrap();
                 UdpFramed::new(socket, BytesCodec::new())
                     .map_err(|e| error!("error reading line: {:?}", e))
-                    .map(|(bytes, _addr)| bytes)
-                    .forward(tx.sink_map_err(|e| error!("error sending event: {:?}", e)))
-                    .map(|_| ())
-            }),
-        );
-
-        rt.spawn(receiver);
-        let _ = rt.block_on(sender).unwrap();
-
-        let messages = rt.block_on(collect_n(rx, 1)).ok().unwrap();
-        assert_eq!(
-            messages[0],
-            Bytes::from("vector.counter:1.5|c|#empty_tag:,normal_tag:value,true_tag\nvector.histogram:2|h|@0.01")
-        );
+                    .map_ok(|(bytes, _addr)| bytes.freeze())
+                    .forward(
+                        tx.sink_compat()
+                            .sink_map_err(|e| error!("error sending event: {:?}", e)),
+                    )
+                    .await
+                    .unwrap()
+            });
+
+            // Add a delay to the write side to let the read side
+            // poll for read interest. Otherwise, this could cause
+            // a race condition in noisy environments.
+            delay_for(Duration::from_millis(100)).await;
```

Comment:
This delay is new, wasn't there before. I wonder if we could get rid of it if we do socket bind on the main test fn, and move it into the `tokio::spawn` instead of doing it in there? As I understand, the only possible new race here is with the socket not being bound when we start collecting messages from it.

Reply:
This delay is not new: https://github.com/timberio/vector/pull/3095/files#diff-42fa092ff1c47ae33a665f6f0df08c4fL371-L376
I think we can move `bind` out of `spawn` and remove delay, hope that we remember what we should change in future if this test start fail :grin:

Reply:
The link to the source doesn't open for me :frowning_face: Which file is it?

Reply:
It's the same file (`src/sinks/statsd.rs`), was removed from lines 371-376